spec/ndx-microscopy.extensions.yaml

```diff
@@ -1,8 +1,260 @@
+  # It would be removed when this structure is merged to core
+  - neurodata_type_def: MicroscopyOpticalChannel
+    neurodata_type_inc: LabMetaData  # Would prefer basic NWBContainer
+    doc: An optical channel used to filter light emission from an imaging space.
+    datasets:
+    - name: description
+      doc: Description or other notes about the channel.
+      dtype: text
+    attributes:
+    - name: indicator
+      doc: Identifier for the indicator pertaining to this optical channel.
+      dtype: text
+    - name: filter_description
+      doc: Metadata information about the filter used by this optical channel.
+      dtype: text
+      required: false
+    - name: emission_wavelength_in_nm
+      doc: Emission wavelength for this optical channel, in nanometers.
+      dtype: numeric
+      required: false
+
+  - neurodata_type_def: ImagingSpace
+    neurodata_type_inc: LabMetaData  # Would prefer basic NWBContainer
+    doc: Metadata about the region of physical space that imaging data was recorded from.
+    datasets:
+    - name: description
+      dtype: text
+      doc: Description of the imaging space.
+    - name: origin_coordinates
+      dtype: float64
+      dims:
+        - - x, y, z
+      shape:
+        - - 3
+      doc: Physical location in stereotactic coordinates for the first element of the grid.
+        See reference_frame to determine what the coordinates are relative to (e.g., bregma).
+      quantity: '?'
+      attributes:
+        - name: unit
+          dtype: text
+          default_value: micrometers
+          doc: Measurement units for origin coordinates. The default value is 'micrometers'.
+    attributes:
+    - name: location
+      dtype: text
+      doc: General estimate of location in the brain being subset by this space.
+        Specify the area, layer, etc.
+        Use standard atlas names for anatomical regions when possible.
+        Specify 'whole-brain' if the entire brain is strictly contained within the space.
+      required: false
+    links:
+    - name: microscope
+      target_type: Microscope
+      doc: Link to Microscope object which contains metadata about the device which imaged this space.
+
+  - neurodata_type_def: PlanarImagingSpace
+    neurodata_type_inc: ImagingSpace
+    doc: Metadata about the 2-dimensional slice of physical space that imaging data was recorded from.
+    datasets:
+    - name: grid_spacing
+      dtype: float64
+      dims:
+      - - x, y
+      shape:
+      - - 2
+      doc: Amount of space between pixels in the specified unit.
+        Specify 'z' only when imaging volume is a regular grid; otherwise only specify 'x' and 'y'.
+        See origin_coordinates to determine where the grid begins.
+      quantity: '?'
+# TODO: deal with grid_spacing units
+#      attributes:
+#        - name: unit
+#          dtype: text
+#          default_value: micrometers
+#          doc: Measurement units for grid spacing. The default value is 'micrometers'.
```

Comment:
BTW @rly 

If I uncomment this part it errors out at `io.write`

```python
    @docval({"name": "spec", "type": Spec, "doc": "the spec to get the attribute value for"},
            {"name": "container", "type": AbstractContainer, "doc": "the container to get the attribute value from"},
            {"name": "manager", "type": BuildManager, "doc": "the BuildManager used for managing this build"},
            returns='the value of the attribute')
    def get_attr_value(self, **kwargs):
        ''' Get the value of the attribute corresponding to this spec from the given container '''
        spec, container, manager = getargs('spec', 'container', 'manager', kwargs)
        attr_name = self.get_attribute(spec)
        if attr_name is None:
            return None
        attr_val = self.__get_override_attr(attr_name, container, manager)
        if attr_val is None:
            try:
>               attr_val = getattr(container, attr_name)
E               AttributeError: 'PlanarImagingSpace' object has no attribute 'grid_spacing__unit'
```

Any idea what's going on there?

What's especially odd is the `origin_coordinates` follows the same exact pattern but works fine

Reply:
Thanks for the ping. This is an issue with HDMF that we haven't encountered before. I filed an issue here https://github.com/hdmf-dev/hdmf/issues/1121 and created a fix here https://github.com/hdmf-dev/hdmf/pull/1122.

I'll aim for a quick bugfix release early this week. Then `grid_spacing__unit` should work.

That said, I was thinking about the "unit" attribute for the `ProbeModel.planar_contour` argument in ndx-extracellular-channels. That is a set of coordinates that needs units too. I removed the ability of the user to set the unit, and changed `planar_contour` to `planar_contour_in_um`. I think there is little value in having the user set the unit. As per some of our recent discussions at the dev hackathon and before, data producers often just want to be told what to do to standardize their data, and software tools and data reusers have a hard time working with data that is too variable across datasets, e.g., when the names of the same kinds of objects differ across datasets for no good reason.

It's probably good to have the `unit` field and fix it to `micrometers`. By fixing it to that value, we can add `_in_um` to the name for extra clarity.

Just my two cents!

Reply:
I think I agree - let's do this in a small follow-up